### PR TITLE
FEATURE: Add "get or create" functionality to TaxonomyTerm

### PR DESCRIPTION
**Pull request by d-rivera-c**
New static function that tries to find a taxonomy term based on an array, and if it doesn't find it, creates one and returns it.

In a lot of Migration cases I needed this, so I think is better to create a function that adding an extension every time extension to handle that. When using that function you always get a taxonomy term back and you don't have to check if it exists.

Added a simple test for the function, checks all 3 cases (argument error, get an existing TT, create a new one).

### CODE REVIEW

**Comment by chillu** [comment]
We could technically have one of those methods on every `DataObject` implementation which has a unique property to filter for. Rails has this as `find_or_create_by_<field>` in ActiveRecord: https://github.com/rails/rails/blob/4-2-stable/guides/source/active_record_querying.md#find_or_create_by

I understand your desire to keep DRY here, but I think this code is better suited in a custom implementation which uses this combination of "find or create" repeatedly. Also, the diverging arguments for filter and create are a bit confusing - as a casual reader, you would think they're the same. In the case of Taxonomy, uniqueness isn't enforced at all. Some use cases won't need uniqueness, others will want more specific uniqueness such as ParentID+Name. Hence I would suggest a single combined argument for both "find" and "create".

I'd love to see this method in SilverStripe core, but I'm a bit reluctant to place it in a module and potentially have inconsistent implementations between modules. @tractorcow What do you think?

File: code/TaxonomyTerm.php [hunk 1]
```diff
@@ -81,6 +81,31 @@ public function getTaxonomy() {
...
+	 * @return TaxonomyTerm found/created toxonomy
+	 */
+	public static function get_or_create($find, $create) {
+		if(!is_array($find) || !is_array($create)) {
+			throw new InvalidArgumentException('Incorrect arguments passed to TaxonomyTerm::get_or_create()');
+		}
+
+		$taxonomy = TaxonomyTerm::get()->filter($find)->first();
+		if(!$taxonomy || !$taxonomy->exists()) {
+			$taxonomy = new TaxonomyTerm($create);
```
Comment (dhensby) [inline code/TaxonomyTerm.php]: `TaxonomyTerm::create()` should be used here for extensibility

File: tests/TaxonomyTermTest.php [hunk 2]
```diff
@@ -43,4 +43,22 @@ public function testSorting() {
...
+	 * Tests correct arguments, get and create a taxonomy
+	 */
+	public function testGetOrCreate() {
+		try {
+			$taxonomy = TaxonomyTerm::get_or_create(1, 'string');
+			$this->fail('An expected exception has not been raised.');
+		} catch (InvalidArgumentException $expected) {}
+		
+		$fruit = TaxonomyTerm::get_or_create(array('Name' => 'Fruit'),
+											 array('Name' => 'Fruit', 'ParentID' => 0));
```
Comment (dhensby) [inline tests/TaxonomyTermTest.php]: Can you sort out the indentation here pleaes

File: tests/TaxonomyTermTest.php [hunk 3]
```diff
@@ -43,4 +43,22 @@ public function testSorting() {
 		$this->assertEquals($fruit->ID, $plant->Children()->first()->ID);
 		$this->assertEquals($vegetable->ID, $plant->Children()->last()->ID);
 	}
+	
+	/**
+	 * Tests correct arguments, get and create a taxonomy
+	 */
+	public function testGetOrCreate() {
```
Comment (dhensby) [inline tests/TaxonomyTermTest.php]: This test seems a bit rudimentary. You should try to get a term that you know exists and assert that it's ID matches the ID of the existing term.

**Comment by chillu** [comment]
Closing this PR for now, I don't think a module-specific `get_or_create()` is leading us down the right path for the reasons outlined above.

**Pull request closed without merging**